src/utils/validations/component/validations/with-validation.hoc.js

```diff
@@ -0,0 +1,188 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import validationsContext from './validations.context';
+import Icon from '../../../../components/icon';
+import validator from '../../validator';
+
+const withValidation = (WrappedComponent) => {
+  class WithValidation extends React.Component {
+    static contextType = validationsContext;
+
+    static propTypes = {
+      children: PropTypes.node,
+      name: PropTypes.string,
+      value: PropTypes.string,
+      validate: PropTypes.func,
+      validations: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      warnings: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      info: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      onBlur: PropTypes.func
+    };
+
+    static defaultProps = {
+      validations: [],
+      warnings: [],
+      info: []
+    }
+
+    static contextTypes = {
+      addInput: PropTypes.func,
+      removeInput: PropTypes.func,
+      adjustCount: PropTypes.func
+    }
+
+    state = {
+      hasError: false,
+      hasWarning: false,
+      hasInfo: false
+    };
+
+    currentFailedValidations = { error: [], warning: [], info: [] };
+
+    componentDidMount = () => {
+      if (this.context && this.context.addInput && this.props.validations.length) {
+        this.context.addInput(this.props.name, this.validate);
+      }
+    }
+
+    componentWillUnmount = () => {
+      if (this.context && this.context.removeInput && this.props.validations.length) {
+        this.context.removeInput(this.props.name);
+      }
+    }
+
+    addFailedValidation = (type, message) => {
+      const { adjustCount } = this.context;
+      this.currentFailedValidations[type].push(message);
+      adjustCount(type, 1);
+    }
+
+    removeFailedValidations = (type) => {
+      const array = this.currentFailedValidations[type];
+      const { adjustCount } = this.context;
```

Comment:
Maybe a better approach is to just to treat the counts on the input rather than to try and count the number of failures per input? Asynchronous validations won't really allow and accurate aggregation to be calculated as the Promise.all will just reject on the first validation that fails.